File: frontend/src/types/mcpServer.ts
/**
 * MCP服务器相关类型定义
 * @description 定义MCP服务器的配置、状态和操作相关类型
 */

/** MCP服务器传输类型 */
export type McpTransportType = 'stdio' | 'sse' | 'streamable-http'

/** MCP服务器实体 */
export interface McpServer {
  /** 主键ID */
  id: number
  /** 服务器名称 */
  name: string
  /** 服务器描述 */
  description?: string
  /** 传输类型 */
  transport: McpTransportType
  /** 是否启用 */
  is_enabled: number
  /** 保持连接：0=调用后释放，1=保持连接 */
  keep_alive: number
  /** 最后连接时间 */
  last_connected_at?: string
  /** 最后错误信息 */
  last_error?: string
  /** 创建时间 */
  create_time?: string
}

/** MCP服务器配置（用于stdio传输） */
export interface McpServerStdioConfig {
  /** 执行命令 */
  command?: string
  /** 命令参数 */
  args?: string[]
  /** 环境变量 */
  env?: Record<string, string>
}

/** MCP服务器配置（用于网络传输） */
export interface McpServerNetworkConfig {
  /** 服务器URL */
  url?: string
  /** 请求头 */
  headers?: Record<string, string>
  /** 工具调用超时时间（秒），1-600 */
  timeout?: number
}

/** MCP服务器完整配置 */
export type McpServerConfig = McpServerStdioConfig & McpServerNetworkConfig

/** 创建MCP服务器参数 */
export interface McpServerCreate {
  /** 服务器名称 */
  name: string
  /** 服务器描述 */
  description?: string
  /** 传输类型 */
  transport: McpTransportType
  /** 是否启用 */
  is_enabled?: number
  /** 保持连接：0=调用后释放，1=保持连接 */
  keep_alive?: number
  /** 服务器配置 */
  config?: McpServerConfig
}

/** 更新MCP服务器参数 */
export interface McpServerUpdate {
  /** 主键ID */
  id: number
  /** 服务器名称 */
  name?: string
  /** 服务器描述 */
  description?: string
  /** 传输类型 */
  transport?: McpTransportType
  /** 是否启用 */
  is_enabled?: number
  /** 保持连接：0=调用后释放，1=保持连接 */
  keep_alive?: number
  /** 服务器配置 */
  config?: McpServerConfig
}

/** MCP工具信息 */
export interface McpToolInfo {
  /** 工具名称 */
  name: string
  /** 工具描述 */
  description?: string
  /** 输入参数Schema */
  input_schema?: Record<string, unknown>
  /** 是否启用 */
  is_enabled?: number
}

/** MCP服务器测试结果 */
export interface McpServerTestResult {
  /** 是否成功 */
  success: boolean
  /** 可用工具列表 */
  tools: McpToolInfo[]
  /** 错误信息 */
  error?: string
}

File: app/models/mcp_server.py
"""
MCP服务器配置模型
"""

from enum import Enum
from datetime import datetime
from typing import Optional
from sqlalchemy import String, Text, SmallInteger, Integer, JSON, DateTime
from sqlalchemy.orm import Mapped, mapped_column
from app.models.base_model import DbBaseModel


class McpTransportType(str, Enum):
    """MCP传输类型"""

    STDIO = "stdio"
    SSE = "sse"
    STREAMABLE_HTTP = "streamable-http"


class McpServer(DbBaseModel):
    """
    MCP服务器配置表模型

    存储MCP服务器的连接配置信息
    """

    __tablename__ = "mcp_server"

    name: Mapped[str] = mapped_column(String(100), nullable=False, comment="服务器名称")
    description: Mapped[Optional[str]] = mapped_column(
        Text, nullable=True, comment="服务器描述"
    )
    transport: Mapped[str] = mapped_column(
        String(20), nullable=False, comment="传输类型：stdio/sse/streamable-http"
    )

    is_enabled: Mapped[int] = mapped_column(
        SmallInteger, nullable=False, default=1, comment="是否启用：0=禁用，1=启用"
    )
    keep_alive: Mapped[int] = mapped_column(
        SmallInteger,
        nullable=False,
        default=1,
        comment="保持连接：0=调用后释放，1=保持连接",
    )
    last_connected_at: Mapped[Optional[datetime]] = mapped_column(
        DateTime, nullable=True, comment="最后连接时间"
    )
    last_error: Mapped[Optional[str]] = mapped_column(
        Text, nullable=True, comment="最后一次连接错误"
    )

    def __repr__(self) -> str:
        return (
            f"<McpServer(id={self.id}, name={self.name}, transport={self.transport})>"
        )


class McpServerConfig(DbBaseModel):
    """
    MCP服务器详细配置表模型

    存储不同传输类型的详细配置
    """

    __tablename__ = "mcp_server_config"

    server_id: Mapped[int] = mapped_column(
        Integer, nullable=False, index=True, comment="关联MCP服务器ID"
    )

    config_key: Mapped[str] = mapped_column(
        String(100), nullable=False, comment="配置键"
    )
    config_value: Mapped[Optional[str]] = mapped_column(
        Text, nullable=True, comment="配置值"
    )

    def __repr__(self) -> str:
        return f"<McpServerConfig(server_id={self.server_id}, key={self.config_key})>"


class McpToolCache(DbBaseModel):
    """
    MCP工具缓存表模型

    缓存从MCP服务器获取的工具列表
    """

    __tablename__ = "mcp_tool_cache"

    server_id: Mapped[int] = mapped_column(
        Integer, nullable=False, index=True, comment="关联MCP服务器ID"
    )

    tool_name: Mapped[str] = mapped_column(
        String(100), nullable=False, comment="工具名称"
    )
    description: Mapped[Optional[str]] = mapped_column(
        Text, nullable=True, comment="工具描述"
    )
    tool_schema: Mapped[Optional[dict]] = mapped_column(
        JSON, nullable=True, comment="工具JSON Schema"
    )

    is_enabled: Mapped[int] = mapped_column(
        SmallInteger, nullable=False, default=1, comment="是否启用：0=禁用，1=启用"
    )

    cached_at: Mapped[datetime] = mapped_column(
        DateTime, default=datetime.now, nullable=False, comment="缓存时间"
    )

    def __repr__(self) -> str:
        return f"<McpToolCache(server_id={self.server_id}, tool={self.tool_name})>"
